tonic/src/codec/decode.rs

```diff
@@ -180,6 +180,7 @@ impl<T> Streaming<T> {
                 }
             };
             let len = self.buf.get_u32() as usize;
+            self.buf.reserve(len);
```

Comment:
Would this be a vector for attack? Since a user could pass any number here and reserve as much memory as possible even if they can't send that on the wire?

Reply:
Yes and no. As far as I see, even before this PR, there was also no protection against the user accumulating an arbitrary large buffer (by, well, just actually sending it). So in this way, this PR doesn't make the situation worse, except that it lowers the attacker's effort. 

On macOS and Linux, allocating huge buffers doesn't actually fault & reserve the corresponding pages until they are actually written to ([overcommit](https://en.wikipedia.org/wiki/Memory_overcommitment)), so the situation doesn't change at all. On Windows, only the size of the page file can be overcommitted, so this would crash immediately on very large values.

Unless I'm wrong and this is actually verified but I just didn't see where (which is very much possible!), I'd argue: yes, this clearly is an issue, but one that should probably be discussed and fixed in a dedicated issue?

Reply:
Thanks for the great write up. Yeah, I think this makes sense then!